## Part 1: The Doomsday Algorithm

The Doomsday algorithm, devised by mathematician J. H. Conway, computes the day of the week any given date fell on. The algorithm is designed to be simple enough to memorize and use for mental calculation.

__Example.__ With the algorithm, we can compute that July 4, 1776 (the day the United States declared independence from Great Britain) was a Thursday.

The algorithm is based on the fact that for any year, several dates always fall on the same day of the week, called the <em style="color:#F00">doomsday</em> for the year. These dates include 4/4, 6/6, 8/8, 10/10, and 12/12.

__Example.__ The doomsday for 2016 is Monday, so in 2016 the dates above all fell on Mondays. The doomsday for 2017 is Tuesday, so in 2017 the dates above will all fall on Tuesdays.

The doomsday algorithm has three major steps:

1. Compute the anchor day for the target century.
2. Compute the doomsday for the target year based on the anchor day.
3. Determine the day of week for the target date by counting the number of days to the nearest doomsday.

Each step is explained in detail below.

### The Anchor Day

The doomsday for the first year in a century is called the <em style="color:#F00">anchor day</em> for that century. The anchor day is needed to compute the doomsday for any other year in that century. The anchor day for a century $c$ can be computed with the formula:
$$
a = \bigl( 5 (c \bmod 4) + 2 \bigr) \bmod 7
$$
The result $a$ corresponds to a day of the week, starting with $0$ for Sunday and ending with $6$ for Saturday.

__Note.__ The modulo operation $(x \bmod y)$ finds the remainder after dividing $x$ by $y$. For instance, $12 \bmod 3 = 0$ since the remainder after dividing $12$ by $3$ is $0$. Similarly, $11 \bmod 7 = 4$, since the remainder after dividing $11$ by $7$ is $4$.

__Example.__ Suppose the target year is 1954, so the century is $c = 19$. Plugging this into the formula gives
$$a = \bigl( 5 (19 \bmod 4) + 2 \bigr) \bmod 7 = \bigl( 5(3) + 2 \bigr) \bmod 7 = 3.$$
In other words, the anchor day for 1900-1999 is Wednesday, which is also the doomsday for 1900.

__Exercise 1.1.__ Write a function that accepts a year as input and computes the anchor day for that year's century. The modulo operator `%` and functions in the `math` module may be useful. Document your function with a docstring and test your function for a few different years.  Do this in a new cell below this one.

In [40]:
def AnchorDay(year):
    """
    This AnchorDay function is used to find the anchor day of a century. 
    User has to input a year and the function will first find out what century is
    And find out the anchor day of that century.
    
    Input: year
    
    Output: an integer corresponds to a day of week, 
    starting with 0 for Sunday and ending with 6 for Saturday
    
    Output is the Doomsday of the target year
    """
    
    century = int(str(year)[-4:-2])
    anchor = (5*(century % 4) + 2) % 7
    return anchor

In [41]:
AnchorDay(1954)

3

### The Doomsday

Once the anchor day is known, let $y$ be the last two digits of the target year. Then the doomsday for the target year can be computed with the formula:
$$d = \left(y + \left\lfloor\frac{y}{4}\right\rfloor + a\right) \bmod 7$$
The result $d$ corresponds to a day of the week.

__Note.__ The floor operation $\lfloor x \rfloor$ rounds $x$ down to the nearest integer. For instance, $\lfloor 3.1 \rfloor = 3$ and $\lfloor 3.8 \rfloor = 3$.

__Example.__ Again suppose the target year is 1954. Then the anchor day is $a = 3$, and $y = 54$, so the formula gives
$$
d = \left(54 + \left\lfloor\frac{54}{4}\right\rfloor + 3\right) \bmod 7 = (54 + 13 + 3) \bmod 7 = 0.
$$
Thus the doomsday for 1954 is Sunday.

__Exercise 1.2.__ Write a function that accepts a year as input and computes the doomsday for that year. Your function may need to call the function you wrote in exercise 1.1. Make sure to document and test your function.

In [42]:
def Doomsday(year):
    """
    This function is used to recognize the doomsday of the targeted year.
    When computing doomsday, we need anchor day of that century first. 
    Here we call the function AnchorDays to find the anchor day of the century.
    Then used the formula to find the Doomsday
    
    Input: target year
    Output: an integer corresponds to a day of week, 
    starting with 0 for Sunday and ending with 6 for Saturday
    
    Output is the Doomsday of the target year
    
    """
    a = AnchorDay(year)
    last2digits = int(str(year)[-2:])
    dooms = (last2digits + last2digits/4 + a) % 7
    return dooms

In [43]:
Doomsday(1954)

0

### The Day of Week

The final step in the Doomsday algorithm is to count the number of days between the target date and a nearby doomsday, modulo 7. This gives the day of the week.

Every month has at least one doomsday:
* (regular years) 1/10, 2/28
* (leap years) 1/11, 2/29
* 3/21, 4/4, 5/9, 6/6, 7/11, 8/8, 9/5, 10/10, 11/7, 12/12

__Example.__ Suppose we want to find the day of the week for 7/21/1954. The doomsday for 1954 is Sunday, and a nearby doomsday is 7/11. There are 10 days in July between 7/11 and 7/21. Since $10 \bmod 7 = 3$, the date 7/21/1954 falls 3 days after a Sunday, on a Wednesday.

__Exercise 1.3.__ Write a function to determine the day of the week for a given day, month, and year. Be careful of leap years! Your function should return a string such as "Thursday" rather than a number. As usual, document and test your code.

In [44]:
def DayofWeek(date):
    """
    This function is used to find the day of the week for a specific date.
    In the begining, we have to find out whether it's on regular year or leap year by dividing year to 4. 
    If modulo 0, it's a leap year.
    Then, we need to find out the doomsday of the same month and use that day as a standard.
    Afterward, we can use the difference of two days to determine the day of week of the inputed date.
    
    Input: date in the string format of month/day/year
    Output: day of week
    """
    date = date.split("/")
    date = [int(i) for i in date]
    if (date[2]%4 == 0 and date[2]%100 != 0) or date[2]%400 == 0:
        dooms_everyyear = [11,29,21,4,9,6,11,8,5,10,7,12]
    else:
        dooms_everyyear = [10,28,21,4,9,6,11,8,5,10,7,12]
    nearest = [date[0],dooms_everyyear[date[0]-1]]
    
    weekday = ["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"]
    doomsday_year = Doomsday(date[2])
    day_diff = date[1] - nearest[1]
    weekday_index = (day_diff%7 + doomsday_year)%7
    return weekday[weekday_index]

In [58]:
DayofWeek("7/21/1954")

'Wednesday'

__Exercise 1.4.__ How many times did Friday the 13th occur in the years 1900-1999? Does this number seem to be similar to other centuries?

<p> 
Although Exercise 1.4 only asked us to find out how many times did Firday the 13th occur between two years, for convinence, I wrote a function to find the Friday the 13th between two date. Then I don't have to worte a function again for Exercise 1.5
</p>

In [46]:
def FridayThe13th(start_date,end_date):
    """
    This function is used to find out how many times did Friday the 13th occur between two specific date.
    
    Input: two dates in the string format of "month/day/year"
    
    We can find out how many 13th between this two date. 
    And use the function DayofWeek to find out the date of each 13th.
    Then we can count how many of day of week are Friday
    
    Output: how many times did Friday the 13th occur
    """
    start = [int(i) for i in start_date.split("/")]
    end = [int(i) for i in end_date.split("/")]
    date13 = []
    
    # 13th of each month in first year
    if start[1] <= 13:
        for month in xrange(start[0],13):
            date13.append("{0}/{1}/{2}".format(str(month),"13",str(start[2])))
    else:
        for month in xrange(start[0]+1,13):
            date13.append("{0}/{1}/{2}".format(str(month),"13",str(start[2])))

    for year in xrange(start[2]+1,end[2]):
        for month in xrange(1,13):
            date13.append("{0}/{1}/{2}".format(str(month),"13",str(year)))

    # 13th of each month in the last year
    if end[1] >= 13:
        for month in xrange(1,end[0]+1):
            date13.append("{0}/{1}/{2}".format(str(month),"13",end[2]))
    else:
        for month in xrange(1,end[0]):
            date13.append("{0}/{1}/{2}".format(str(month),"13",end[2]))

    weekdayofthirteen = [DayofWeek(d) for d in date13]
    return weekdayofthirteen.count("Friday")


In [47]:
FridayThe13th("1/1/1900","12/31/1999")

172

There're 172 Friday the 13th in 19th century. Now, let's find out how many time Friday the 13th in 17th and 18th century.

In [48]:
[FridayThe13th("1/1/1800","12/31/1899"),FridayThe13th("1/1/1700","12/31/1799")]

[172, 172]

For every century, there're 172 Friday the 13th. The numbder of Friday the 13th is the same for every century.

__Exercise 1.5.__ How many times did Friday the 13th occur between the year 2000 and today?

In [49]:
FridayThe13th("1/1/2000","1/19/2017")

30

## Part 2: 1978 Birthdays

__Exercise 2.1.__ The file `birthdays.txt` contains the number of births in the United States for each day in 1978. Inspect the file to determine the format. Note that columns are separated by the tab character, which can be entered in Python as `\t`. Write a function that uses iterators and list comprehensions with the string methods `split()` and `strip()` to  convert each line of data to the list format

```Python
[month, day, year, count]
```
The elements of this list should be integers, not strings. The function `read_birthdays` provided below will help you load the file.

In [50]:
def read_birthdays(file_path):
    """Read the contents of the birthdays file into a string.
    
    Arguments:
        file_path (string): The path to the birthdays file.
        
    Returns:
        string: The contents of the birthdays file.
    """
    with open(file_path) as file:
        return file.read()

In [51]:
def BDStringToList(text):
    """
    This function is used to convert each line of data to the list format [month, day, year, count]
    
    Input: text read through the function read_birthdays
    
    Output: a list of list containing the contents of the birthdays file in the format of [month, day, year, count]
    """
    text = text.strip()
    text = text.split("\n")
    bd_list = []
    for item in text:
        bd_count = item.split()
        date =  bd_count[0].split("/")
        date[2] = "19" + date[2]
        bd = date + [bd_count[1]]
        bd = [int(a) for a in bd]
        bd_list = bd_list + [bd]
    return bd_list

In [52]:
birthday = BDStringToList(read_birthdays("/Users/alice/STA141B/Homework1/birthdays.txt"))
birthday[0:5]

[[1, 1, 1978, 7701],
 [1, 2, 1978, 7527],
 [1, 3, 1978, 8825],
 [1, 4, 1978, 8859],
 [1, 5, 1978, 9043]]

__Exercise 2.2.__ Which month had the most births in 1978? Which day of the week had the most births? Which day of the week had the fewest? What conclusions can you draw? You may find the `Counter` class in the `collections` module useful.

In here, I first created a list containing 12 elements and each element = 0. Then I used month-1 as the index and keep sum the counts into the list in that index. In the end, I use index and max function to find out which month has the most counts in 1978.

In [53]:
count_month = [0]*12
for item in birthday:
    count_month[item[0]-1] = count_month[item[0]-1] + item[3]
count_month.index(max(count_month))+1

8

August had the most births in 1978.

For the day of week, I also use the same method. However, to find out the day of week, we have to use the function DayofWeek I wrote before. Therefore, I first turn the list into a string in a format of "month/day/year". Then we can find out the the day of week and sum the births up.

In [54]:
for date in birthday:
    date_string = "{0}/{1}/{2}".format(date[0],date[1],date[2])
    date = date.append(DayofWeek(date_string))

weekdaydic = ["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"]
weekdaycount = [0]*7
for date in birthday:
    weekdaycount[weekdaydic.index(date[4])] += date[3]


In [55]:
[weekdaydic[weekdaycount.index(max(weekdaycount))],weekdaydic[weekdaycount.index(min(weekdaycount))]]

['Tuesday', 'Sunday']

Tuesday had the most births and Sunday had the fewest births.

__Exercise 2.3.__ What would be an effective way to present the information in exercise 2.2? You don't need to write any code for this exercise, just discuss what you would do.

I think an effective way to present the information in exercise 2.2 is to turn the list into plot, such as histogram.